General isai

fritzgoebel opened the pull request:
This PR extends ISAI from triangular to general and spd matrices.
When generated for a general or spd matrix, it can be used as a preconditioner on its own, i.e. outside of a factorization.
In the spd case, an ISAI L is computed on the lower triangular half of A, a diagonal scaling step makes the result equivalent to an FSPAI preconditioner, so an approximation on the exact Cholesky factorization. In this case, the preconditioner is no longer a single CSR matrix but a composition of L^T and L.

For rows of size up to 32, the corresponding small dense systems are solved directly, for larger rows an excess system is built up as it is for triangular systems. The excess system is solved using GMRES with a block-Jacobi preconditioner.
A new factory parameter `excess_limit` sets a threshold to the size of the excess system. If set, the excess system gets solved as soon as this threshold is exceeded, the solution scattered to the result matrix and a new excess system is built up for the remaining rows, and so on.

fritzgoebel commented:
format!

upsj commented on core/preconditioner/isai_kernels.hpp:
```diff
@@ -51,29 +51,41 @@ namespace kernels {
-    template <typename ValueType, typename IndexType>                     \
-    GKO_DECLARE_ISAI_GENERATE_EXCESS_SYSTEM_KERNEL(ValueType, IndexType); \
-    template <typename ValueType, typename IndexType>                     \
+        matrix::Csr<ValueType, IndexType> *inverse, const size_type e_start,  \
+        const size_type e_end)
```
nit: `const` is an implementation detail and not part of the interface
```suggestion
        matrix::Csr<ValueType, IndexType> *inverse, size_type e_start,        \
        size_type e_end)
```

upsj commented on core/preconditioner/isai_kernels.hpp:
```diff
@@ -51,29 +51,41 @@ namespace kernels {
         const IndexType *excess_rhs_ptrs, const IndexType *excess_nz_ptrs,   \
         matrix::Csr<ValueType, IndexType> *excess_system,                    \
-        matrix::Dense<ValueType> *excess_rhs)
+        matrix::Dense<ValueType> *excess_rhs, const size_type excess_start,  \
+        const size_type excess_end)
```
```suggestion
        matrix::Dense<ValueType> *excess_rhs, size_type excess_start,        \
        size_type excess_end)
```

upsj commented on include/ginkgo/core/preconditioner/isai.hpp:
```diff
@@ -57,10 +57,10 @@ namespace preconditioner {
  *
- * ISAI can either generate a lower triangular matrix, or an upper triangular
- * matrix.
+ * ISAI can either generate a general matrix, a lower triangular matrix, or an
```
maybe
```suggestion
 * ISAI can either be generated for a general matrix, a lower triangular matrix, or an
```

upsj commented on include/ginkgo/core/preconditioner/isai.hpp:
```diff
@@ -98,15 +98,18 @@ class Isai : public EnableLinOp<Isai<IsaiType, ValueType, IndexType>>,
     using index_type = IndexType;
     using transposed_type =
-        Isai<IsaiType == isai_type::lower ? isai_type::upper : isai_type::lower,
+        Isai<IsaiType == isai_type::general
```
We should note somewhere that ISAI(transpose) != transpose(ISAI)

upsj commented on common/components/warp_blas.hpp.inc:
```diff
@@ -83,6 +83,52 @@ __device__ __forceinline__ void apply_gauss_jordan_transform(
+    bool &__restrict__ status)
+{
+    auto key_col_elem = group.shfl(row[key_col], key_row);
+    auto key_rhs_elem = group.shfl(rhs[0], key_row);
```
Suggestion: Since you only access `rhs[0]`, you could also pass `rhs` by reference instead of by pointer, which saves the `[0]` in this function.

upsj commented on common/preconditioner/isai_kernels.hpp.inc:
```diff
@@ -125,22 +129,26 @@ __forceinline__ __device__ void generic_generate(
                     if (valid) {
-                        trisystem(nz, i_idx) = m_values[m_row_begin + m_idx];
+                        if (m_col_idxs[m_row_begin + m_idx] < row && col == row)
+                            atomic_add(&rhs_one_idxs[row % subwarps_per_block],
+                                       size_type{1});
```
I would suggest you [(sub)warp-aggregate](https://developer.nvidia.com/blog/cuda-pro-tip-optimized-filtering-warp-aggregated-atomics/) the results, then you should be able to avoid using atomics and shared memory:
```cpp
count += popcnt(subwarp.ballot());
```
Alternative: Store partial counts and reduce afterwards
```cpp
local_count += (m_col_idxs[m_row_begin + m_idx] < row && col == row);
// ...
count = reduce(subwarp, local_count)
```

upsj commented on core/preconditioner/isai.cpp:
```diff
@@ -108,17 +114,22 @@ std::shared_ptr<Csr> extend_sparsity(std::shared_ptr<const Executor> &exec,
 void Isai<IsaiType, ValueType, IndexType>::generate_inverse(
-    std::shared_ptr<const LinOp> input, bool skip_sorting, int power)
+    std::shared_ptr<const LinOp> input, bool skip_sorting, int power,
+    int excess_limit)
```
```suggestion
    IndexType excess_limit)
```

upsj commented on core/preconditioner/isai.cpp:
```diff
@@ -127,38 +138,86 @@ void Isai<IsaiType, ValueType, IndexType>::generate_inverse(
+    if (total_excess_dim > 0) {
+        bool done = false;
+        size_type block = 0;
+        while (true) {
+            // build the excess sparse triangular system
+            size_type excess_dim = 0;
+            size_type excess_start = block;
+            const auto block_offset = exec->copy_val_to_host(
+                excess_block_ptrs.get_const_data() + block);
+            const auto nnz_offset = exec->copy_val_to_host(
+                excess_row_ptrs_full.get_const_data() + block);
+            while (excess_dim < excess_lim && block < num_rows) {
+                block++;
+                excess_dim = exec->copy_val_to_host(
+                                 excess_block_ptrs.get_const_data() + block) -
+                             block_offset;
+            }
+            if (excess_dim == 0) break;
```
We do a lot of individual copies here, which might not be too efficient. I would suggest that we instead create a host-side copy of excess_block_ptrs and use std::upper_bound to find the end of the current batch of blocks. Then we could also write this in a nicer loop:
```cpp
for (size_type excess_start = 0; excess_start < total_excess_dim; excess_start = find_next_block_index_lambda(excess_start))
```

upsj commented on core/preconditioner/isai.cpp:
```diff
@@ -127,38 +138,86 @@ void Isai<IsaiType, ValueType, IndexType>::generate_inverse(
+                trs_factory =
+                    Gmres::build()
+                        .with_preconditioner(
+                            bj::build().with_max_block_size(32u).on(exec))
```
Two comments here:
1. we could make the solver factory a factory parameter of ISAI that defaults to this choice, then people can also use GMRES with BJ for triangular systems if they like, similar to what the ILU preconditioner does.
2. Since we already have block structure information, why not pass this on to the preconditioner as block pointers clamped at 32? That might significantly improve the preconditioner quality.

upsj commented on omp/preconditioner/isai_kernels.cpp:
```diff
@@ -146,12 +151,14 @@ void generic_generate(std::shared_ptr<const DefaultExecutor> exec,
                         m_cols + m_begin, m_size, i_cols + i_begin, i_size,
                         [&](IndexType, IndexType m_idx, IndexType i_idx) {
-                            trisystem(i, i_idx) = m_vals[m_idx + m_begin];
+                            if (m_cols[m_idx + m_begin] < row && col == row)
+                                rhs_one_idx++;
```
nit: formatting
```suggestion
                            if (m_cols[m_idx + m_begin] < row && col == row) {
                                rhs_one_idx++;
                            }
```

upsj commented on reference/preconditioner/isai_kernels.cpp:
```diff
@@ -138,12 +142,14 @@ void generic_generate(std::shared_ptr<const DefaultExecutor> exec,
                     m_cols + m_begin, m_size, i_cols + i_begin, i_size,
                     [&](IndexType, IndexType m_idx, IndexType i_idx) {
-                        trisystem(i, i_idx) = m_vals[m_idx + m_begin];
+                        if (m_cols[m_idx + m_begin] < row && col == row)
+                            rhs_one_idx++;
```
nit: formatting
```suggestion
                        if (m_cols[m_idx + m_begin] < row && col == row) {
                            rhs_one_idx++;
                        }
```

upsj commented on omp/preconditioner/isai_kernels.cpp:
```diff
@@ -240,14 +247,111 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+            }
+
+            // solve transposed system
+            std::vector<IndexType> perm(size, {});
```
nit: this is allocated outside of our control, you could use gko::vector instead. But it looks to me like we don't need the vector anyways.

The funny thing is, when I saw your reference implementation, I immediately thought: "This could maybe be improved with a permutation array", so nice 😆

upsj commented on omp/preconditioner/isai_kernels.cpp:
```diff
@@ -240,14 +247,111 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+                                 transposed_system + col * (size + 1), size) +
+                    col;
+                swap_rows(col, row, size, transposed_system, size);
+                swap(perm[row], perm[col]);
```
You don't seem to be using perm anywhere?

upsj reviewed (approved):
LGTM - great job! I didn't look over the (especially reference) tests in too much detail, but the rest looked solid.

hartwiganzt commented on hip/preconditioner/isai_kernels.hip.cpp:
```diff
@@ -168,6 +168,24 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
 
 
+template <typename ValueType, typename IndexType>
+void scale_excess_solution(std::shared_ptr<const DefaultExecutor>,
```
Do you anywhere document your strategy, like what you do with the long-row systems?

yhmtsai commented on omp/preconditioner/isai_kernels.cpp:
```diff
@@ -240,14 +248,115 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+            rhs[rhs_one_idx] = one<ValueType>();
+
+            // fill in transposed system
+            ValueType *transposed_system = new ValueType[size * size];
```
Miss the deconstructor. Can it use the gko::Array?
gko::matrix can not be used here due to circular dependence. Does it also apply to gko::Array?

upsj replied:
Good catch! No, you can use gko::Array, since it is a header-only type. I preallocated temporary storage for the triangular solve, you should be able to use the same approach here.

yhmtsai commented on omp/preconditioner/isai_kernels.cpp:
```diff
@@ -240,14 +248,115 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+                transposed_system[col * size + col] = one<ValueType>() / d;
+            }
+
+            if (spd) {
+                const auto scal = one<ValueType>() / sqrt(rhs[size - 1]);
+                for (IndexType row = 0; row < size; row++) {
+                    rhs[row] *= scal;
+                }
+            }
```
omp seem to miss the spd cases test.

yhmtsai reviewed (changes requested):
Miss free memory in one lambda function

upsj commented on cuda/test/preconditioner/isai_kernels.cpp:
```diff
@@ -102,6 +112,29 @@ class Isai : public ::testing::Test {
+        std::ifstream mtxstream{std::string{gko::matrices::location_isai_mtxs} +
+                                name};
+        auto result = gko::read<ReadMtx>(mtxstream, ref);
+        // to avoid removing 0s, the matrices store 12345 instead
+        for (gko::size_type i = 0; i < result->get_num_stored_elements(); ++i) {
+            auto &val = result->get_values()[i];
+            if (val == static_cast<value_type>(12345.0)) {
+                val = 0;
+            }
+        }
```
I'm to blame for this, I think I wanted to make the rows longer artificially by inserting numerical zeros into the sparsity pattern, but read() usually removes them.

fritzgoebel commented on reference/test/preconditioner/isai_kernels.cpp:
```diff
@@ -614,14 +856,123 @@ TYPED_TEST(Isai, KernelGenerateExcessULongrow)
+    GKO_ASSERT_MTX_EQ_SPARSITY(result, this->spd_csr_longrow_inv_partial);
+    GKO_ASSERT_MTX_NEAR(result, this->spd_csr_longrow_inv_partial,
+                        r<value_type>::value);
+    // no row above the size limit -> zero array
```
```suggestion
```

fritzgoebel commented on include/ginkgo/core/preconditioner/isai.hpp:
```diff
@@ -57,22 +57,30 @@ namespace preconditioner {
+ * ISAI(A)^T != ISAI(A^T).
  *
  * For more details on the algorithm, see the paper
  * <a href="https://doi.org/10.1016/j.parco.2017.10.003">
```
No, I think we can add the europar paper here later?

upsj commented on reference/preconditioner/isai_kernels.cpp:
```diff
@@ -34,6 +34,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 
 #include <algorithm>
+#include <iostream>
```
```suggestion
```

upsj commented on reference/preconditioner/isai_kernels.cpp:
```diff
@@ -228,13 +236,115 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+            rhs[rhs_one_idx] = one<ValueType>();
+
+            // fill in transposed system
+            ValueType *transposed_system = new ValueType[size * size];
```
Same as OpenMP: You should use pre-allocated storage here

upsj commented on reference/preconditioner/isai_kernels.cpp:
```diff
@@ -228,13 +236,115 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+            ValueType *transposed_system = new ValueType[size * size];
+            for (auto row = 0; row < size; row++) {
+                for (auto col = 0; col < size; col++) {
+                    transposed_system[row * size + col] = system(col, row);
```
How about using a range for transposed_system as well?

upsj commented on reference/test/preconditioner/isai_kernels.cpp:
```diff
@@ -154,6 +219,16 @@ class Isai : public ::testing::Test {
         u_csr_longrow_e_rhs = read<Dense>("isai_u_excess_rhs.mtx");
         u_csr_longrow_inv_partial = read<Csr>("isai_u_inv_partial.mtx");
         u_csr_longrow_inv = read<Csr>("isai_u_inv.mtx");
+        a_csr_longrow = read<Csr>("isai_a.mtx");
+        a_csr_longrow_e = read<Csr>("isai_a_excess.mtx");
+        a_csr_longrow_e_rhs = read<Dense>("isai_a_excess_rhs.mtx");
+        a_csr_longrow_inv_partial = read<Csr>("isai_a_inv_partial.mtx");
+        a_csr_longrow_inv = read<Csr>("isai_a_inv.mtx");
+        spd_csr_longrow = read<Csr>("isai_spd.mtx");
+        spd_csr_longrow_e = read<Csr>("isai_spd_excess.mtx");
+        spd_csr_longrow_e_rhs = read<Dense>("isai_spd_excess_rhs.mtx");
+        spd_csr_longrow_inv_partial = read<Csr>("isai_spd_inv_partial.mtx");
+        spd_csr_longrow_inv = read<Csr>("isai_spd_inv.mtx");
```
We may want to think about ways to programatically generate these matrices?

upsj commented on include/ginkgo/core/preconditioner/isai.hpp:
```diff
@@ -57,22 +57,30 @@ namespace preconditioner {
- * ISAI can either generate a lower triangular matrix, or an upper triangular
- * matrix.
+ * ISAI can either be generated for a general square matrix, a lower triangular
+ * matrix, an upper triangular matrix or a spd matrix.
```
from the pronounciation, I would assume this is similar to `an essential`
```suggestion
 * matrix, an upper triangular matrix or an spd matrix.
```

upsj commented on include/ginkgo/core/preconditioner/isai.hpp:
```diff
@@ -84,8 +92,9 @@ enum struct isai_type { lower, upper };
- * @tparam IsaiType  determines if the ISAI is generated for a lower triangular
- *                   matrix or an upper triangular matrix
+ * @tparam IsaiType  determines if the ISAI is generated for a general square
+ * matrix, a lower triangular matrix, an upper triangular matrix or an spd
```
nit: formatting

upsj commented on include/ginkgo/core/preconditioner/isai.hpp:
```diff
@@ -143,6 +164,25 @@ class Isai : public EnableLinOp<Isai<IsaiType, ValueType, IndexType>>,
+         * solution has been computed.
+         * Must be at least 0, default value 0.
+         */
+        index_type GKO_FACTORY_PARAMETER_SCALAR(excess_limit, 0);
```
I would prefer size_type, but this will hopefully become a signed type in the future, which would require a check for negative values as well.

upsj commented on hip/test/preconditioner/isai_kernels.hip.cpp:
```diff
@@ -53,7 +53,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 
-enum struct matrix_type { lower, upper };
+enum struct matrix_type { lower, upper, general };
```
spd

upsj commented on cuda/test/preconditioner/isai_kernels.cpp:
```diff
@@ -41,19 +41,22 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 
-enum struct matrix_type { lower, upper };
+enum struct matrix_type { lower, upper, general };
```
spd

upsj commented on common/preconditioner/isai_kernels.hpp.inc:
```diff
@@ -233,6 +239,55 @@ __global__ __launch_bounds__(default_block_size) void generate_u_inverse(
+
+        ValueType sol = subwarp.shfl(rhs, perm);
+
+        if (spd) {
+            sol /= sqrt(subwarp.shfl(rhs, num_elems - 1));
+        }
```
suggestion for clarity:
```suggestion
        if (spd) {
            auto diag = subwarp.shfl(rhs, num_elems - 1);
            sol /= sqrt(diag);
        }
```

upsj commented on common/preconditioner/isai_kernels.hpp.inc:
```diff
@@ -283,36 +343,63 @@ __global__ __launch_bounds__(default_block_size) void generate_excess_system(
+    const IndexType offset = excess_block_ptrs[e_start];
+    const IndexType block_begin = excess_block_ptrs[warp_id] - offset;
+    const IndexType block_end = excess_block_ptrs[warp_id + 1] - offset;
+    const ValueType scal =
```
same here

upsj commented on core/preconditioner/isai.cpp:
```diff
@@ -108,17 +119,49 @@ std::shared_ptr<Csr> extend_sparsity(std::shared_ptr<const Executor> &exec,
+        // Init arrays
+        Array<IndexType> inverted_col_idxs{exec, inverted_nnz};
+        Array<ValueType> inverted_vals{exec, inverted_nnz};
+        std::shared_ptr<Csr> inverted_base = Csr::create(
```
you could also use `auto` and `share(...)` here

upsj commented on core/preconditioner/isai.cpp:
```diff
@@ -127,38 +170,94 @@ void Isai<IsaiType, ValueType, IndexType>::generate_inverse(
+    if (total_excess_dim > 0) {
+        bool done = false;
+        size_type block = 0;
+        while (true) {
```
can we write this in terms of a normal `while` loop instead? it seems the condition you are checking here is `while(block < num_rows)`

upsj reviewed (changes requested):
LGTM in general, only 3 points:

1. Do you test the excess_limit behavior somewhere? It looks to me like all tests use `0, num_rows` as the limit parameters in generate_excess_system.
2. Did you try out explicitly specifying the block pointers for Block Jacobi?
3. Could you try addressing the concerns about memory movement in the excess block pointers for excess limit handling?

fritzgoebel commented:
> LGTM in general, only 3 points:
> 
>     1. Do you test the excess_limit behavior somewhere? It looks to me like all tests use `0, num_rows` as the limit parameters in generate_excess_system.
> 
>     2. Did you try out explicitly specifying the block pointers for Block Jacobi?
> 
>     3. Could you try addressing the concerns about memory movement in the excess block pointers for excess limit handling?

1: good catch!
2: Not yet, I plan on doing that in a different pr later.
3: Please check if that's okay :)

thoasm reviewed (approved):
LGTM!
Some small nits, but nothing major.

upsj commented on common/components/warp_blas.hpp.inc:
```diff
@@ -83,6 +83,52 @@ __device__ __forceinline__ void apply_gauss_jordan_transform(
+    auto key_col_elem = group.shfl(row[key_col], key_row);
+    auto key_rhs_elem = group.shfl(rhs[0], key_row);
+    if (key_col_elem == zero<ValueType>()) {
+        // TODO: implement error handling for GPUs to be able to properly
+        //       report it here
+        status = false;
```
For reference, that comment and `status` was already there 2 years ago, so I think Goran meant more general error handling mechanisms.

thoasm replied:
Ah, good to know, thanks. In that case I will mark it as resolved.

fritzgoebel replied:
I don't think so, I carried this over from `apply_gauss_jordan_transform` and there it is also still present..

yhmtsai commented on include/ginkgo/core/preconditioner/isai.hpp:
```diff
@@ -143,6 +164,25 @@ class Isai : public EnableLinOp<Isai<IsaiType, ValueType, IndexType>>,
+        size_type GKO_FACTORY_PARAMETER_SCALAR(excess_limit, 0u);
+
+        /**
+         * Factory for the Excess System solver
```
Could you also add the default solver for different cases if it is not set by user

yhmtsai commented on omp/test/preconditioner/isai_kernels.cpp:
```diff
@@ -52,7 +52,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-enum struct matrix_type { lower, upper };
+enum struct matrix_type { lower, upper, general, spd };
 class Isai : public ::testing::Test {
```
```suggestion


class Isai : public ::testing::Test {
```

yhmtsai commented on cuda/test/preconditioner/isai_kernels.cpp:
```diff
@@ -41,19 +41,22 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-enum struct matrix_type { lower, upper };
+enum struct matrix_type { lower, upper, general, spd };
 class Isai : public ::testing::Test {
```
```suggestion


class Isai : public ::testing::Test {
```

yhmtsai commented on hip/test/preconditioner/isai_kernels.hip.cpp:
```diff
@@ -53,7 +53,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-enum struct matrix_type { lower, upper };
+enum struct matrix_type { lower, upper, general, spd };
 class Isai : public ::testing::Test {
```
```suggestion


class Isai : public ::testing::Test {
```

yhmtsai reviewed (approved):
LGTM. some minor format and documentation suggestion
